John Sullivan, Cole, Jared
CSC310
2/19/2020
Assignment03

In [2]:
import pandas as pd
from sklearn import tree
from treeviz import tree_print
from sklearn.metrics import accuracy_score

# This is our first dataset, the provided tennis dataset. 

In [4]:
#Necessary for reading the file, and then we're just going to peek at what's inside of it. 
tennis = pd.read_csv("tennis_numeric.csv")
tennis.head()

outlook_sunny  outlook_overcast  outlook_rainy  temperature  humidity  \
0              1                 0              0           85        85   
1              1                 0              0           80        90   
2              0                 1              0           83        86   
3              0                 0              1           70        96   
4              0                 0              1           68        80   

   windy play  
0      0   no  
1      1   no  
2      0  yes  
3      0  yes  
4      0  yes

In [5]:
tennisFeatures = tennis.drop(['play'], axis=1)
tennisFeatures.head()

outlook_sunny  outlook_overcast  outlook_rainy  temperature  humidity  \
0              1                 0              0           85        85   
1              1                 0              0           80        90   
2              0                 1              0           83        86   
3              0                 0              1           70        96   
4              0                 0              1           68        80   

   windy  
0      0  
1      1  
2      0  
3      0  
4      0

In [6]:
targetDataFrame = pd.DataFrame(tennis['play'])
targetDataFrame.head()

play
0   no
1   no
2  yes
3  yes
4  yes

In [7]:
decisionTree = tree.DecisionTreeClassifier(criterion='entropy')
decisionTree.fit(tennisFeatures, targetDataFrame)
tree_print(decisionTree, tennisFeatures)

if outlook_overcast =< 0.5: 
  |then if humidity =< 82.5: 
  |  |then if temperature =< 66.5: 
  |  |  |then no
  |  |  |else yes
  |  |else if humidity =< 95.5: 
  |  |  |then no
  |  |  |else yes
  |else yes
<------->
Tree Depth:  3


In [8]:
prediction = decisionTree.predict(tennisFeatures)      
labels = pd.DataFrame(prediction)  
labels.columns = ['play']
labels.equals(targetDataFrame)

True

In [10]:
print("Model Accuracy for the Tennis data set: {}".format(accuracy_score(targetDataFrame, prediction)))

Model Accuracy for the Tennis data set: 1.0


As you can see above, the tree we've created models the dataset with 100% accuracy. 

# This is our second data set, the Caesarian dataset .

In [13]:
#import the file 
caesarian = pd.read_csv("caesarian.csv")
caesarian.head()

Age  Delivery number  Delivery time  Blood of Pressure  Heart Problem  \
0   22                1              0                  2              0   
1   26                2              0                  1              0   
2   26                2              1                  1              0   
3   28                1              0                  2              0   
4   22                2              0                  1              0   

   Caesarian  
0      False  
1       True  
2      False  
3      False  
4       True

In [7]:
c_features = caesarian.drop(['Caesarian'], axis=1)
c_features.head()

Age  Delivery number  Delivery time  Blood of Pressure  Heart Problem
0   22                1              0                  2              0
1   26                2              0                  1              0
2   26                2              1                  1              0
3   28                1              0                  2              0
4   22                2              0                  1              0

In [8]:
targetDataFrame = pd.DataFrame(caesarian['Caesarian'])
targetDataFrame.head()

Caesarian
0      False
1       True
2      False
3      False
4       True

In [9]:
decisionTree = tree.DecisionTreeClassifier(criterion='entropy')
decisionTree.fit(c_features, targetDataFrame)
tree_print(decisionTree, c_features)

if Heart Problem =< 0.5: 
  |then if Age =< 21.5: 
  |  |then if Age =< 18.5: 
  |  |  |then if Age =< 17.5: 
  |  |  |  |then True
  |  |  |  |else False
  |  |  |else True
  |  |else if Delivery number =< 2.5: 
  |  |  |then if Age =< 31.5: 
  |  |  |  |then if Age =< 29.5: 
  |  |  |  |  |then if Age =< 28.5: 
  |  |  |  |  |  |then if Age =< 27.5: 
  |  |  |  |  |  |  |then if Delivery time =< 0.5: 
  |  |  |  |  |  |  |  |then if Age =< 26.5: 
  |  |  |  |  |  |  |  |  |then if Delivery number =< 1.5: 
  |  |  |  |  |  |  |  |  |  |then if Blood of Pressure =< 0.5: 
  |  |  |  |  |  |  |  |  |  |  |then False
  |  |  |  |  |  |  |  |  |  |  |else if Age =< 23.5: 
  |  |  |  |  |  |  |  |  |  |  |  |then if Blood of Pressure =< 1.5: 
  |  |  |  |  |  |  |  |  |  |  |  |  |then True
  |  |  |  |  |  |  |  |  |  |  |  |  |else False
  |  |  |  |  |  |  |  |  |  |  |  |else if Blood of Pressure =< 1.5: 
  |  |  |  |  |  |  |  |  |  |  |  |  |then False
  |  |  |  |  |  |  |  |  |  |  

We originally thought that this decision tree would be pretty far off, since it looks so strange, but if you look below you can see the model accuracy is actually almost spot on. 

In [10]:
prediction = decisionTree.predict(c_features)      
labels = pd.DataFrame(prediction)  
labels.columns = ['Caesarian']
labels.equals(targetDataFrame)

False

In [12]:
("Model Accuracy for the Caesarian data set: {}".format(accuracy_score(targetDataFrame, prediction)))

'Model Accuracy for the Caesarian data set: 0.9615384615384616'

As you can see above, the model accuracy isn't too far off to our surprise. We thought because the decision tree looked so obscure and strange, that our model accuracy may be way off. But it appears that it's not too bad with a solid 96% accuracy. It's not totally accurate, but it's something!